cpp/cmake_modules/DefineOptions.cmake

```diff
@@ -233,6 +233,8 @@ Note that this requires linking Boost statically" OFF)
 
   define_option(ARROW_WITH_ZLIB "Build with zlib compression" ON)
 
+  define_option(ARROW_REQUIRE_ENCRYPTION "Fail if OpenSSL is not found" OFF)
```

Comment:
should this be `PARQUET_REQUIRE_ENCRYPTION` and be added to the PARQUET option list?

Reply:
never mind, I see that this is used for `ARROW_FLIGHT` as well.

Reply:
I think we should use the previous `define_option(PARQUET_BUILD_ENCRYPTION "Build Parquet with encryption support" ON)`. We already have the other flag `ARROW_FLIGHT` where encryption is required.
The build logic will be:
1. Turn OFF PARQUET_BUILD_ENCRYPTION if ARROW_PARQUET is OFF
2. Error if either `PARQUET_BUILD_ENCRYPTION` or `ARROW_FLIGHT` is ON and OpenSSL is not found.
3. By default enable PARQUET_BUILD_ENCRYPTION if OpenSSL is available and ARROW_PARQUET is ON

```
if (PARQUET_BUILD_ENCRYPTION AND NOT ARROW_PARQUET)
  set(PARQUET_BUILD_ENCRYPTION OFF)
endif()
if(PARQUET_BUILD_ENCRYPTION OR ARROW_FLIGHT)
  # This must work
  find_package(OpenSSL REQUIRED)
  set(ARROW_USE_OPENSSL ON)
elseif(ARROW_PARQUET)
  # Enable Parquet encryption if OpenSSL is there, but don't fail if it's not
  find_package(OpenSSL QUIET)
  CMAKE_DEPENDENT_OPTION(ARROW_USE_OPENSSL "Build with OpenSSL support" ON
                         "OPENSSL_FOUND" OFF)
  # use this flag to include the relevant C++ encryption files for build
  CMAKE_DEPENDENT_OPTION(PARQUET_BUILD_ENCRYPTION "Build with OpenSSL support" ON
                         "OPENSSL_FOUND" OFF)
endif()
```

Reply:
IIUC what you're proposing is equivalent to the status quo because the default of `PARQUET_BUILD_ENCRYPTION` is `ON`. So if I build with Parquet support but OpenSSL isn't found, the build fails. That's the behavior I was trying to remedy here, and why I introduced a new flag with `REQUIRE` in the name to make explicit that if you turn it on, it must work (but it's `OFF` by default).

Reply:
Ah! yes! How about setting the default `PARQUET_BUILD_ENCRYPTION` to `OFF` in my proposal?

Reply:
My intention here is to have a CMake variable with a name that describes the parquet encryption support.